src/mlpack/methods/ann/vanilla_rbm_impl.hpp

```diff
@@ -0,0 +1,201 @@
+}
+
+ /* 
+  * This function calculates
+  * the free energy of the model
+  */
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    FreeEnergy(arma::mat&& input)
+{
+  arma::mat output;
+  visible.Forward(std::move(input), std::move(output));
+  SoftplusFunction::Fn(output, output);
+  return  -arma::accu(output) -arma::dot(input, visible.Bias());
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Evaluate(const arma::mat& /* parameters*/, const size_t i)
+{
+  
+    double freeEnergy, freeEnergyChain;
+    arma::mat negativeSample;
+    arma::vec currentInput = predictors.unsafe_col(i);
+    // Do not use unsafe col predictor as we change the input
+    freeEnergy = FreeEnergy(std::move(currentInput));
+    Gibbs(currentInput, std::move(negativeSample));
+    freeEnergyChain = FreeEnergy(std::move(negativeSample));
+
+    return arma::mean(freeEnergy) - arma::mean(freeEnergyChain);
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    SampleHidden(arma::mat&& input, arma::mat&& output)
+{
+  visible.Sample(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    SampleVisible(arma::mat&& input, arma::mat&& output)
+{
+  hidden.Sample(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Gibbs(arma::mat input, arma::mat&& output)
+{
+  if (persistence && !state.is_empty())
+    input = state;
+
+  for (size_t j = 0; j < numSteps; j++)
+  {
+    // Use probabilties for updation till the last step(section 3 hinton)
+    SampleHidden(std::move(input), std::move(output));
+    SampleVisible(std::move(output), std::move(input));
+  }
+  output = input;
+  state = input;
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Gradient(const size_t input, arma::mat& output)
+{
+  arma::mat positiveGradient, negativeGradient, negativeSamples;
+  // Collect the positive gradients
+  CalcGradient(std::move(predictors.col(input)), std::move(positiveGradient));
+
+  // Collect the negative samples
+  Gibbs(std::move(predictors.col(input)), std::move(negativeSamples));
+
+  // Collect the negative gradients
+  CalcGradient(std::move(negativeSamples), std::move(negativeGradient));
+
+  output = positiveGradient - negativeGradient;
```

Comment:
This is the main issue: the sign is invalid.
So write the following
```
output = -(positiveGradient - negativeGradient);
```
Or change the signs of `positiveGradient` and `negativeGradient`.
If you calculate the derivatives of the energy function you will see that they has got the minus sign in the beginning.